p2p/client.go

```diff
@@ -2,135 +2,228 @@ package p2p
+	}
+	err := c.host.Close()
+	if err != nil {
+		c.logger.Error("failed to close P2P host", "error", err)
+		return err
+	}
+	return dhtErr
+}
+
 func (c *Client) listen() error {
-	// TODO(tzdybal): consider requiring listen address in multiaddress format
-	maddr, err := GetMultiAddr(c.conf.ListenAddress)
+	var err error
+	maddr, err := multiaddr.NewMultiaddr(c.conf.ListenAddress)
 	if err != nil {
 		return err
 	}
 
-	//TODO(tzdybal): think about per-client context
-	host, err := libp2p.New(context.Background(), libp2p.ListenAddrs(maddr), libp2p.Identity(c.privKey))
+	c.host, err = libp2p.New(c.ctx,
+		libp2p.ListenAddrs(maddr),
+		libp2p.Identity(c.privKey),
+	)
 	if err != nil {
 		return err
 	}
-	for _, a := range host.Addrs() {
-		c.logger.Info("listening on", "address", a)
+
+	for _, a := range c.host.Addrs() {
+		c.logger.Info("listening on", "address", fmt.Sprintf("%s/p2p/%s", a, c.host.ID()))
 	}
 
-	c.host = host
 	return nil
 }
 
-func (c *Client) bootstrap() error {
-	if len(strings.TrimSpace(c.conf.Seeds)) == 0 {
+func (c *Client) setupDHT() error {
+	seedNodes := c.getSeedAddrInfo(c.conf.Seeds)
+	if len(seedNodes) == 0 {
 		c.logger.Info("no seed nodes - only listening for connections")
-		return nil
 	}
-	seeds := strings.Split(c.conf.Seeds, ",")
+
+	for _, sa := range seedNodes {
+		c.logger.Debug("seed node", "addr", sa)
+	}
+
+	var err error
+	c.dht, err = dht.New(c.ctx, c.host, dht.Mode(dht.ModeServer), dht.BootstrapPeers(seedNodes...))
+	if err != nil {
+		return fmt.Errorf("failed to create DHT: %w", err)
+	}
+
+	err = c.dht.Bootstrap(c.ctx)
+	if err != nil {
+		return fmt.Errorf("failed to bootstrap DHT: %w", err)
+	}
+	return nil
+}
+
+func (c *Client) peerDiscovery() error {
+	err := c.setupPeerDiscovery()
+	if err != nil {
+		return err
+	}
+
+	err = c.advertise()
+	if err != nil {
+		return err
+	}
+
+	err = c.findPeers()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Client) setupPeerDiscovery() error {
+	// wait for DHT
+	<-c.dht.RefreshRoutingTable()
+	c.disc = discovery.NewRoutingDiscovery(c.dht)
+	return nil
+}
+
+func (c *Client) advertise() error {
+	// TODO(tzdybal): add configuration parameter for re-advertise frequency
+	discovery.Advertise(c.ctx, c.disc, c.getNamespace(), discovery.TTL(1*time.Hour))
+	return nil
+}
+
+func (c *Client) findPeers() error {
+	// TODO(tzdybal): add configuration parameter for max peers
+	peerCh, err := c.disc.FindPeers(c.ctx, c.getNamespace(), cdiscovery.Limit(60))
+	if err != nil {
+		return err
+	}
+
+	for peer := range peerCh {
+		go c.tryConnect(peer)
+	}
+
+	return nil
+}
+
+// tryConnect attempts to connect to a peer and logs error if necessary
+func (c *Client) tryConnect(peer peer.AddrInfo) {
+	err := c.host.Connect(c.ctx, peer)
+	if err != nil {
+		c.logger.Error("failed to connect to peer", "peer", peer, "error", err)
+	}
+}
+
+func (c *Client) getSeedAddrInfo(seedStr string) []peer.AddrInfo {
```

Comment:
cosmos-sdk is providing a single, comma-separated string anyways, so converting from string to slice is required.
But it may fit better in the `conv` package, as this logic is not client-specific.

Reply:
After re-thinking I'll leave it as is. I tried to:
1. only split string into slice in conv; but it makes code more complicated and actual parsing is located in two packages
2. move entire method to conv; but error handling is client-specific - I can't think of other use of "parse seed nodes and only log errors if necessary, without returning errors" function ;)

Reply:
Can you pls provide an example of such a string with multiple seeds? How is that related to Cosmos? I was thinking that this is a multiaddr string

Reply:
cosmos-sdk app, will use optimint as a drop-in replacement of tendermint (see [master-optimint](https://github.com/lazyledger/cosmos-sdk/blob/master-optimint/server/start.go#L265) branch on our fork).
For better user experience, we reuse configurations, command line options, etc from cosmos-sdk. And cosmos-sdk apps have a config/option `Seeds` which is a comma separated string. The actual change is that we expect multiaddr format, not cosmos-sdk format.
Example is: `/ip4/192.168.1.1/tcp/1234/p2p/seed1_id,/ip4/192.168.1.2/tcp/4321/p2p/seed2_id`
Ref: https://github.com/lazyledger/optimint/issues/16